### PR TITLE
Add array::count, array::range, array::shorter and array::longer rules

### PR DESCRIPTION
**Pull request by vladgorenkin**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ✔️ updated error messages for number::lower and number::higher
| New feature?  | ✔️

Allows to validate number of elements in array

### CODE REVIEW

File: src/Validation/src/Checker/ArrayChecker.php [hunk 1]
```diff
@@ -31,4 +38,27 @@ public function of($value, $checker): bool
 
         return true;
     }
+
+    public function count($value, ?int $min = null, ?int $max = null, ?int $exact = null): bool
```
Comment (SerafimArts) [inline src/Validation/src/Checker/ArrayChecker.php:42]: Shouldn't `$min` be the default 0?

Comment (vladgorenkin) [reply src/Validation/src/Checker/ArrayChecker.php]: FIxed, thanks

**Comment by SerafimArts** [comment]
Maybe it makes sense then to remove "exact" and change the "max" logic to `$max ??= $min`?

Then it will turn out:
```php
$checker->count($arr, 4); // Strictly count($arr) === 4
$checker->count($arr, 1, 42); // In range from 1 to 42
```

What do you think?

**Comment by vladgorenkin** [comment]
And `$checker->count($arr, 1, null)` to only specify lower limit?

**Comment by SerafimArts** [comment]
> And $checker->count($arr, 1, null) to only specify lower limit?

What about `$checker->count($arr, 1, INF)`?

**Comment by vladgorenkin** [comment]
>  what about a more semantically consistent variant `$checker->count($arr, 1, INF)`?

\INF is problematic because it is special PHP value which cannot be exported without additional processing. 

```php
var_dump(json_encode(['array::count', 1, \INF]));
```
gives me `bool(false)`.

**Comment by vladgorenkin** [comment]
It is also important to take into account that this function will mostly be used through validation DSL, where you can specify arguments names:
```php
['array::count', 'exact' => 4]
['array::count', 'min' => 2]
['array::count', 1, 5]
```

**Comment by SerafimArts** [comment]
> \INF is problematic because it is special PHP value which cannot be exported without additional processing.

It is not entirely clear why these values should be converted into json?

**Comment by vladgorenkin** [comment]
> > \INF is problematic because it is special PHP value which cannot be exported without additional processing.
> 
> It is not entirely clear why these values should be converted into json?

I see at least 2 possible use cases:
- to share validation rules with frontend
- to store in external storage. For instance, dynamic form with validation stored in DB

All existing rules are json-encodable, but this one isn't. I understand this might not be a priority for you, but I'm just sharing my concerns.

**Comment by vladgorenkin** [comment]
We could also use `count($arr, 2, 2)` for exact case (as Symfony's count constraint does) but it looks ugly to me

**Comment by SerafimArts** [comment]
> I see at least 2 possible use cases:

Yea, agree. I think this cases may well happen.

> We could also use count($arr, 2, 2) for exact case (as Symfony's count constraint does) but it looks ugly to me

Well, this option is better than 3 different arguments.

---

Or make two different validators? One with range and the other with a strict amount?

```
array::range($arr, 1); // 1+
array::range($arr, 1, 42); // 1 ... 42

// and

array::count($arr, 42); // 42
```

**Comment by vladgorenkin** [comment]
I would say `range` is a bit unclear name but it's more consistent taking in mind existing `string::range` and `number::range`. Speaking of consistency: do we want to create something like `string::shorter/longer` but for arrays?

**Comment by SerafimArts** [comment]
> I would say range is a bit unclear name

But it is also located under the "namespace" `array`, so there shouldn't be any problems with a misunderstanding of what this validator is doing. Isnt it?

**Comment by vladgorenkin** [comment]
> > I would say range is a bit unclear name
> 
> But it is also located under the "namespace" `array`, so there shouldn't be any problems with a misunderstanding of what this validator is doing.

Well, my first association with `array::range(1, 42)` is a list of sequent number from 1 to 42 :) But ok I dont' mind. I'll update PR tomorrow

**Comment by SerafimArts** [comment]
> Well, my first association with `array::range(1, 42)` is a list of sequent number from 1 to 42 :) 

What about `array::in_range(1, 42)`?

**Comment by vladgorenkin** [comment]
> > Well, my first association with `array::range(1, 42)` is a list of sequent number from 1 to 42 :)
> 
> What about `array::in_range(1, 42)`?

`range` is better due to constistency.
I noticed a problem with `string::shorter/longer` and `number::lower/higher` -- they are actually `lower or equal` and `higher or equal`. What's worse is that validation messages do not reflect this either

**Review by SerafimArts — APPROVED** [review]

**Comment by vladgorenkin** [comment]
Should I also check if number of elements in `shorter/longer` is equal to `$length` as it's done in other checkers? And should I adjust error messages in all checkers accordingly (unfortunately it's going to break translations for users)?

**Comment by wolfy-j** [comment]
Yes it makes sense to adjust errors for all the checkers but mention that in PR and next release. Worse case it will fallback for English version.

**Comment by vladgorenkin** [comment]
> Yes it makes sense to adjust errors for all the checkers but mention that in PR and next release. Worse case it will fallback for English version.

✔️

**Review by SerafimArts — APPROVED** [review]

**Pull request merged**